addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,137 @@
+/* global Symbol */
+import Ember from 'ember';
+import { buildSelector, assign } from '../../helpers';
+import { create } from '../../create';
+import { count } from '../count';
+import Ceibo from 'ceibo';
+
+import { throwBetterError } from '../../better-errors';
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._countPage = create({
+      count: count(definition.scope, { testContainer: definition.testContainer })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  objectAt(index) {
+    if (this._items[index] === undefined) {
+      let { definition, parent, key } = this;
+      let scope = buildSelector({}, definition.scope, { at: index });
+
+      let finalizedDefinition = assign({}, definition);
+
+      finalizedDefinition.scope = scope;
+
+      let tree = create(finalizedDefinition, { parent });
+
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${key}(${index})`;
+
+      this._items[index] = tree;
+    }
+
+    return this._items[index];
+  }
+
+  get length() {
+    return this._countPage.count;
+  }
+
+  toArray() {
+    let { length } = this;
+
+    let array = Ember.A();
+
+    for (let i = 0; i < length; i++) {
+      array.push(this.objectAt(i));
+    }
+
+    return array;
+  }
+
+  map(...args) {
+    return this.toArray().map(...args);
+  }
+
+  mapBy(...args) {
+    return this.toArray().mapBy(...args);
+  }
+
+  forEach(...args) {
+    return this.toArray().forEach(...args);
+  }
+
+  findOne(query) {
+    let result = this.findAll(query);
+
+    if (result.length > 1) {
+      throwBetterError(
+        this.parent,
+        this.key,
+        `Expected at most one result from findOne query, but found ${result.length}`
+      );
+    }
+
+    return result[0];
+  }
+
+  findAll(query) {
```

Comment:
Yes, exactly. I debated including `filter` and `filterBy`, but I decided against it because they are pretty redundant, but less safe (no assertions, etc) and not the recommended path. On top of that, if we included them it wouldn't make much sense to exclude other array functions, like `find` (which would be doubly confusing) and others.

Reply:
I'd prefer `filter()` more for the sake of learnability.
How does `findAll` differ from `filter`? It seems they both have the same signatures and do very similar thing.

Reply:
`findAll` can accept an object that is a query, matching lodash's `filter/find` methods. We decided not to name it `filter` since A. it may be confusing to users who expect it to work exactly like native array's `filter`, and B. we may want to add `filter/filterBy` passthroughs at some point in the future.